File: scanner/tests/fixtures/VulnerableApp.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.Statement;
import java.sql.ResultSet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

public class VulnerableApp {

    private Connection getConnection() throws Exception {
        return DriverManager.getConnection("jdbc:mysql://localhost/app", "root", "");
    }

    // SQL Injection
    public void getUser(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String userId = request.getParameter("id");
        Connection conn = getConnection();
        Statement stmt = conn.createStatement();
        ResultSet rs = stmt.executeQuery("SELECT * FROM users WHERE id = " + userId);
    }

    // Command Injection
    public void ping(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String host = request.getParameter("host");
        Runtime.getRuntime().exec("ping -c 1 " + host);
    }

    // Internal call chain
    private String validateInput(String data) {
        if (data == null || data.isEmpty()) {
            throw new IllegalArgumentException("empty");
        }
        return data.trim();
    }

    private ResultSet queryDatabase(String id) throws Exception {
        Connection conn = getConnection();
        Statement stmt = conn.createStatement();
        return stmt.executeQuery("SELECT * FROM users WHERE id = " + id);
    }

    public void getUser2(HttpServletRequest request, HttpServletResponse response) throws Exception {
        String id = validateInput(request.getParameter("id"));
        ResultSet rs = queryDatabase(id);
    }
}

File: scanner/tests/fixtures/VulnerableApp.kt
import java.sql.DriverManager
import java.sql.Connection

class UserController(private val dbUrl: String) {
    fun getUser(userId: String): String {
        val conn = DriverManager.getConnection(dbUrl)
        val stmt = conn.createStatement()
        val query = "SELECT * FROM users WHERE id = '$userId'"
        val rs = stmt.executeQuery(query)
        return rs.getString(1)
    }

    fun deleteUser(userId: String) {
        val conn = DriverManager.getConnection(dbUrl)
        val stmt = conn.createStatement()
        stmt.executeUpdate("DELETE FROM users WHERE id = '$userId'")
    }

    fun runCommand(cmd: String) {
        Runtime.getRuntime().exec(cmd)
    }
}

fun handleRequest(input: String): String {
    val controller = UserController("jdbc:mysql://localhost/db")
    return controller.getUser(input)
}

fun executeCommand(command: String) {
    val controller = UserController("jdbc:mysql://localhost/db")
    controller.runCommand(command)
}
